File: leetcode/easy/111_二叉树的最小深度/111_二叉树的最小深度.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#递归法
class Solution:
    def getDepth(self, node: Optional[TreeNode]) -> int:
        if node is None:
            return 0

        leftDepth = self.getDepth(node.left)
        rightDepth = self.getDepth(node.right)

        if node.left is None and node.right is not None:
            return 1 + rightDepth
        if node.left is not None and node.right is None:
            return 1 + leftDepth

        result = 1 + min(leftDepth, rightDepth)
        return result

    def minDepth(self, root: Optional[TreeNode]) -> int:
        return self.getDepth(root)



#迭代法
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        depth = 0
        que = [root]
        while que:
            size = len(que)
            depth += 1
            flag = 0
            for _ in range(size):
                node = que.pop(0)
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
                if not node.left and not node.right:
                    flag = 1
                    break
            if flag == 1:
                break
        return depth

File: leetcode/easy/111_二叉树的最小深度/111_二叉树的最小深度.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

//递归法
func getDepth(node *TreeNode) int {
	if node == nil {
		return 0
	}

	leftDepth := getDepth(node.Left)
	rightDepth := getDepth(node.Right)

	if node.Left == nil && node.Right != nil {
		return 1 + rightDepth
	}
	if node.Left != nil && node.Right == nil {
		return 1 + leftDepth
	}

	if leftDepth < rightDepth {
		return 1 + leftDepth
	}
	return 1 + rightDepth
}

func minDepth(root *TreeNode) int {
	return getDepth(root)
}



//迭代法
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	depth := 0
	que := []*TreeNode{root}
	for len(que) > 0 {
		size := len(que)
		depth++
		flag := 0
		for i := 0; i < size; i++ {
			node := que[0]
			que = que[1:]
			if node.Left != nil {
				que = append(que, node.Left)
			}
			if node.Right != nil {
				que = append(que, node.Right)
			}
			if node.Left == nil && node.Right == nil {
				flag = 1
				break
			}
		}
		if flag == 1 {
			break
		}
	}
	return depth
}